TINKERPOP-2245 Added UnifiedChannelizer

spmallette opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-2245

The `UnifiedChannelizer` converges the thread pools used by sessions and sessionless requests while also streamlining the processing model for scripts and bytecode. There is a fair bit of explanation and discussion in the dev list thread linked in the JIRA. We will keep the old channelizers as the defaults for now, but the `UnifiedChannelizer` is meant to be the future and will eventually become the default allowing us to remove the `OpProcessor` infrastructure as well as the old channelizers in general. 

VOTE +1

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+            // not active, and hence we should break the loop.
+            if (!nettyContext.channel().isActive()) {
+                if (managedTransactionsForRequest) {
+                    closeTransaction(gremlinContext, Transaction.Status.ROLLBACK);
```
should log at debug since we are silently taking an action without any prompt from client. same case if we are breaking the sending of results.

spmallette replied:
this code is gone now...removed as it is technically covered elsewhere on channel close.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+        final RequestMessage msg = gremlinContext.getRequestMessage();
+        final ChannelHandlerContext nettyContext = gremlinContext.getChannelHandlerContext();
+        final MessageSerializer serializer = nettyContext.channel().attr(StateKey.SERIALIZER).get();
+        final boolean useBinary = nettyContext.channel().attr(StateKey.USE_BINARY).get();
```
do we need this additional attribute? this decision could be made by looking at the serializer which is already stored per channel.

spmallette replied:
You can't really look at the serializer to determine this value as GraphSON message serializers can be either form. There might be a better way to do it than an channel attribute though - not sure. This wasn't new code - I pulled it from the old stuff.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+        }
+    }
+
+    private void processAuditLog(final Settings settings, final ChannelHandlerContext ctx, final Object gremlinToExecute) {
```
This should probably be a method in a class call AuditLogger.

spmallette replied:
that's a different tangle which i dont think we should try to deal with in this PR as it relates to a wholly different feature of 3.5.0:

https://tinkerpop.apache.org/docs/3.5.0-SNAPSHOT/upgrade/#_gremlin_server_audit_logging

divijvaidya replied:
sounds good.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/Rexster.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * The {@code Rexster} interface is essentially a form of "worker", named for Gremlin's trusty canine robot friend
```
Is it a form of "worker" or form of "work"? Since it implements Runnable, it seems like Rexter is a unit of work that is processed by RexterExecutorService.

This implements runnable (which signifies that it is a standalone task) and yet it's implementations have a addTask() method and contains executorService as it's member?

Here's a suggestion, let me know what you think about it:
- GremlinOpExecutor is a service backed which executes a GremlinOpTask. 
- GremlinOp can either be TransactionOp or QueryOp etc.
- GremlinOpExecutor has lifecycle states which begins at server start up and ends on server shutdown.
- GremlinOpTask to GremlinOpExecutor are provided from a Netty Event loop thread pool (which we shouldn't block) and hence, any underlying implementation should delegate to it's own thread pool ASAP.
- Interface SessionExecutor containing methods like setSessionThread etc.
- The concrete implementation, MultiTaskGremlinOpExecutor can execute multiple GremlinOps at the same time (using a queue and worker thread pool underneath) and it also implements SessionExecutor. 
- GremlinOpExecutor should have methods such as:
   - Future executeGremlinOpAsync
   
   
I am suggesting the above model so that we can get some unification in vocabulary used in the code base. Currently we have terms such as task, job, rexter, worker, executor, graph operation etc. sued across the code base without any concrete meaning to them which makes it difficult to read code.

spmallette replied:
ok - I'll not indulge myself this time with fun naming schemes 😄

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/util/RexsterExecutorService.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    protected void beforeExecute(final Thread t, final Runnable r) {
+        if (r instanceof RexsterFutureTask)
```
else? what are the other types of tasks that we are expecting here?

spmallette replied:
i dont think there are other types of tasks at this point and since `beforeExecute()` is a side-effect i didnt think anything else had to happen here.  we just want to look for `RexsterFutureTask` and assign the thread if we can.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+    protected final ConcurrentMap<String, Rexster> sessions;
+    protected final Set<String> aliasesUsedByRexster = new HashSet<>();
+
+    protected enum CloseReason { UNDETERMINED, CHANNEL_CLOSED, SESSION_TIMEOUT, REQUEST_TIMEOUT, NORMAL }
```
Nice!! please document each of them. e.g. what does normal mean? client initiated close before completion or close after completion?

spmallette replied:
i renamed and reworked the names a bit and added docs.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+        this.initialChannel = gremlinContext.getChannelHandlerContext().channel();
+
+        // close Rexster if the channel closes to cleanup and close transactions
+        this.initialChannel.closeFuture().addListener(f -> {
```
When we are iterating for results, we do something similar where we try to see if channel is inactive, then we throw an exception. would closing the rexter over here interrupt the underlying execution as well? do we handle the race condition between the exception I mentioned above vs. close here?

spmallette replied:
that's an interesting connection you've made there. upon reflection, i think the code in `handleIterator()` that you are referring to could be removed. i gave it a try and there were no testing failures. on the other hand, i'm not 100% sure that particular code pathway had tests guaranteed to validate it. it's late on my end - i won't commit/push until morning and had a fresh mind t consider it.

spmallette replied:
ended up sticking with the change and removing the code in `handleIterator()`

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+        this.initialChannel.closeFuture().addListener(f -> {
+            if (closeReason.compareAndSet(null, CloseReason.CHANNEL_CLOSED)) {
+                // cancel session worker or it will keep waiting for items to appear in the session queue
+                cancel(true);
```
cancel is already called from inside close

spmallette replied:
true - i think i got overzealous in closing/cancelling stuff. fixed that.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+    @Override
+    public synchronized void close() {
+        // already closing/closed
+        if (!sessions.containsKey(sessionId)) return;
```
how do we handle this case:
1. request1 for sessionA gets enqueued and starts processing
2. request2 for sessionA gets enqueued and hasn't started processing yet
3. client closes session
4. request1 gets interrupted and rolledback. session ID is removed from the map
5. request2 is dequeued, it tried to find if a sessionA exists and it finds that it doesn't so it creates another one and executes successfully.

From a client perspective, they sent 2 requests for a session and then closed the session which should have rolledback both of them but now client observes that request2 is successful. We can say that request2 will not commit ever unless we get a commit from the client for sessionA but it still ends up using resources to execute.

Instead of a Map, I think we should be using  DS where we can evict the closed sessions after some amount of time i.e. we still keep their IDs in the DS even when closed. If a request arrives for a closed session we throw exception. Since we don't want this DS to increase indefinitely, we should evict closed entries after some time. Ideally this DS should evict closed session after X number of requests have been dequeued where X is the number of requests in queue when the session was closed.

spmallette replied:
If request2 has made it to the queue when request1 was interrupted, then it will be removed from the queue with error returned to the client.  It will not use resources to execute. I think this test is showing the scenario you're talking about.

https://github.com/apache/tinkerpop/commit/04a3b9a9db2008be82e4593f5b9ae4a16a4d4b81

spmallette replied:
> Instead of a Map, I think we should be using DS where we can evict the closed sessions after some amount of time i.e. we still keep their IDs in the DS even when closed. If a request arrives for a closed session we throw exception. Since we don't want this DS to increase indefinitely, we should evict closed entries after some time. Ideally this DS should evict closed session after X number of requests have been dequeued where X is the number of requests in queue when the session was closed.

maybe my mind is blanking but what is a "DS"? i'm going to feel dumb when you say what that is but i considered some kind of system like the one you describe to keep the ids but then it wouldn't let us replicate the original functionality of sessions where the id could be re-used, which would prevent applications from working the way they always have. i wasnt sure if that was a step we should be taking.

divijvaidya reviewed (commented):
Initial review and didn't go into depth yet. The main comments are about the code structure and in the next review I will check for edge cases and functionality.

Thanks for picking up this long due task!

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/GremlinServer.java:
```diff
@@ -284,8 +285,10 @@ private static Channelizer createChannelizer(final Settings settings) throws Exc
-                if (gremlinExecutorService != null)
-                    gremlinExecutorService.awaitTermination(30000, TimeUnit.MILLISECONDS);
+                if (gremlinExecutorService != null) {
+                    if (!gremlinExecutorService.awaitTermination(30000, TimeUnit.MILLISECONDS))
```
we don't typically do that in TP unless the single line is hard to read. as this case is nested perhaps it meets that criteria, so i'll change it.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/channel/UnifiedChannelizer.java:
```diff
@@ -0,0 +1,77 @@
+ * A {@link Channelizer} that supports websocket and HTTP requests and does so with the most streamlined processing
+ * model for Gremlin Server introduced with 3.5.0.
+ */
+public class UnifiedChannelizer extends AbstractChannelizer {
```
i had it in my mind that we would be getting rid of this zoo of channelizers which is why i broke with the old naming. will we really have a "V2" or will we simply deprecate away and remove those old channelizers.  if we added http2 support would it be "Http2Channelizer" or "Http2AndHttpAndWebsocketChannelizer"? Just seems like that naming where we include the protocol in the name no longer makes sense. There will just be one `Channelizer` and that's the `UnifiedChannelizer` and most users will never even need to care to know that the `Channelizer` notion even exists. Do you envision something different here?

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+        return new SimpleBindings(graphManager.getAsBindings());
+    }
+
+    protected Bindings mergeBindingsFromRequest(final Context gremlinContext, final Bindings bindings) throws RexsterException {
```
we've always had bindings/parameters for anything that needed scripts - definitely not removed or that would cause script performance to tank as the scriptengine cache would have no effect.

divijvaidya replied:
ack. resolving comment.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+    /**
+     * If {@link Bytecode} is detected to contain a "graph operation" then it gets processed by this method.
+     */
+    protected void handleGraphOperation(final Context gremlinContext, final Bytecode bytecode, final Graph graph) throws Exception {
```
I just had "graph operations" as static members of `Bytecode` but had thought about making an enum out of them at one point when i came up with them. just sorta forgot about it. i'll add them as a `GraphOp` enum and see what happens

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+                final Map<String, Object> attributes = generateStatusAttributes(gremlinContext,
+                        ResponseStatusCode.NO_CONTENT, Collections.emptyIterator());
+                gremlinContext.writeAndFlush(ResponseMessage.build(msg)
+                            .code(ResponseStatusCode.NO_CONTENT)
```
`NO_CONTENT` is "success" as it is a 200 level message that means "it worked but there were no results for the client to worry about" - https://github.com/apache/tinkerpop/blob/c1235268a5ec92444c0885ff460c02e705a04666/gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/message/ResponseStatusCode.java#L41-L46

divijvaidya replied:
ack. resolving comment.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractRexster.java:
```diff
@@ -0,0 +1,758 @@
+                // write back a no-op for success
+                final Map<String, Object> attributes = generateStatusAttributes(gremlinContext,
+                        ResponseStatusCode.NO_CONTENT, Collections.emptyIterator());
+                gremlinContext.writeAndFlush(ResponseMessage.build(msg)
```
A failure here would get handled the same way as those in `handleIterator()` - the exception would get converted to a `RexterException` in `handleException()` and then handled centrally here: https://github.com/apache/tinkerpop/blob/b4776a56d1f955edd4704d75c95d11419b280925/gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java#L141-L167 

Looking at it now, I suppose we could offer a specific warn as part of a try/catch/rethrow for this case if you think it would be helpful to have a specific log message for.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+ */
+public class MultiRexster extends AbstractRexster {
+    private static final Logger logger = LoggerFactory.getLogger(MultiRexster.class);
+    protected final BlockingQueue<Context> queue = new LinkedBlockingQueue<>();
```
I guess i was just matching what we already allowed? We could bound this queue and the general request queue - super old issue: https://issues.apache.org/jira/browse/TINKERPOP-709 and start rate limiting basically?

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Override
+    public boolean acceptingTasks() {
```
that's fair...changing

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+
+    @Override
+    public void addTask(final Context gremlinContext) {
+        if (acceptingTasks())
```
Changing to:

```java
    @Override
    public boolean addTask(final Context gremlinContext) {
        return isAcceptingTasks() && queue.offer(gremlinContext);
    }
```

that doesn't let us easily differentiate a rate limiting issue from a "this session is just closed" issue, if we did https://issues.apache.org/jira/browse/TINKERPOP-709 but maybe we don't worry on that too much for now.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Override
+    public void close() {
```
I'll add the javadoc. The call is mostly async as it basically just interrupts the thread running the `Rexster` instance and it is that gremlinPool thread that finishes its own cleanup. if this method was called by the gremlinPool thread it would be pretty fast to finish because there's nothing heavy left to do.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Override
+    public void close() {
```
i tried to explain further in the javadoc and in the previous comment.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    private void cancelRequestTimeout() {
+        if (requestCancelFuture != null && !requestCancelFuture.isDone())
```
debug is fine - adding

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiRexster.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    private void stopAcceptingRequests() {
+        ending.set(true);
```
that makes sense. i think there was a time i was struggling sorting out what was setting that thing first so i just made it `true` no matter what.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/channel/UnifiedChannelizer.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public void finalize(final ChannelPipeline pipeline) {
+        super.finalize(pipeline);
+        pipeline.remove(PIPELINE_OP_SELECTOR);
```
done - added a fatty explanation

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/WsAndHttpChannelizerHandler.java:
```diff
@@ -66,11 +66,11 @@ public void channelRead(final ChannelHandlerContext ctx, final Object obj) {
                 final ChannelHandler authenticator = pipeline.get(PIPELINE_AUTHENTICATOR);
                 pipeline.remove(PIPELINE_AUTHENTICATOR);
-                pipeline.addAfter(PIPELINE_HTTP_RESPONSE_ENCODER, PIPELINE_AUTHENTICATOR, authenticator);
+                pipeline.addAfter(PIPELINE_HTTP_AGGREGATOR, PIPELINE_AUTHENTICATOR, authenticator);
```
documentation is present now for the before/after of pipeline

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/Rexster.java:
```diff
@@ -0,0 +1,89 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.server.handler;
```
I think there was some package level access going on between handler and these classes.  we already have other classes in "handler" like `Frame` and `HttpHandlerUtil` and more that support handlers so i think it's ok.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/Rexster.java:
```diff
@@ -0,0 +1,89 @@
+ * the {@link SingleRexster} and the {@link MultiRexster} should handle most cases quite well and are extensible for
+ * providers.
+ */
+public interface Rexster extends Runnable {
```
generally accepting your naming suggestions from here: https://github.com/apache/tinkerpop/pull/1414/files#r608709781

spmallette commented:
I've done a lot of work here and have largely addressed all your comments (hopefully). A few big ticket things to pay attention to:

1. Despite my earlier reply, I went ahead and implemented rate limiting from https://issues.apache.org/jira/browse/TINKERPOP-709
2. I renamed almost everything and which helped me to realized that i could kill some code which ended up getting rid of some "tasks" and "executors" and simplified some things. Basic structure now is that the `UnifiedHandler` (still not sure i want to go with the functional naming you suggested) creates a `SessionTask` which is submitted to a `Session` to be executed. The `Session` consumes a thread from the `gremlinPool` to do its work and is either a `SingleTaskSession` (meaning it executes the `SessionTask` that called for its creation and then completes) or is a `MultiTaskSession` (meaning it executes the initial `SessionTask` that called for its creation and then holds the thread awaiting more `SessionTask` instances to be submitted to it).
3. Lots more javadoc that i hope is more clear and a lot more inline comments.

Well, done with this for the weekend unless i'm struck by some fresh bit of thinking. I kinda like where it is. I think you're initial review helped refine things.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiTaskSession.java:
```diff
@@ -0,0 +1,283 @@
+            throw new IllegalStateException(String.format("Worker has no initial context for session: %s", getSessionId()));
+
+        try {
+            startTransaction(sessionTask);
```
I don't completely understand this. Wouldn't this start a new trx for every request in the same session? Is that expected behaviour?

spmallette replied:
the answer to this might be more clear given my improved comments here: https://github.com/apache/tinkerpop/pull/1414/files#r611785656 

but basically, that method is only called when the session is started.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractSession.java:
```diff
@@ -0,0 +1,819 @@
+     */
+    protected void startTransaction(final SessionTask sessionTask) {
+        // check if transactions are open and rollback first to ensure a fresh start.
+        graphManager.rollbackAll();
```
if transactions are open, it means we have a bug in our code where we are failing to properly close transactions. please add a warn log here. I am actually in favor of throwing an exception and not processing this request at all because an invalid transaction state is a serious bug which can impact correctness of data.

spmallette replied:
agreed to throw an exception. we've always been sorta relaxed about handling open transactions and that was probably a mistake. quietly rolling back just hides the problem. good idea.

divijvaidya commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiTaskSession.java:
```diff
@@ -0,0 +1,283 @@
+    public void run() {
+        // allow the Session to know about the thread that is running it - the thread really only has relevance
+        // once the session has started.
+        this.sessionThread = Thread.currentThread();
```
How do we ensure that all requests in the same session are run using the same thread (since trx might be thread bound)?

spmallette replied:
Adding a comment as follows to further clarify:

```java
        // allow the Session to know about the thread that is running it which will be a thread from the gremlinPool
        // the thread really only has relevance once the session has started. this thread is then held below in the
        // while() loop awaiting items to arrive on the request queue. it will hold until the session is properly
        // closed by the client or through interruption/error
```

divijvaidya replied:
ah, in that case, if the default worker pool size is 8, then, we would not be able to process 9th new concurrent session? Because all the threads will be stuck serving the existing sessions? This will also impact single-session aka session-less requests.

I understand that this is not different from current behaviour but today a single session request does not get impacted by multi session query because they share different worker thread pools. We are changing that assumption here. A customer who is relying on the behaviour above to run longer/multiple queries on sessions and low latency faster query on session-less will see a regression from this change. We should document that somewhere.

In future, we should fix this limitation. This is the "starvation by session requests" I was referring to in my email reply over the dev mailing list. There are multiple ways to fix this but perhaps we should treat it as out of scope for this PR.

spmallette replied:
so, the basic problem is that the TinkerPop API doesn't really let us move a transaction context across threads. There is a notion of the "threaded transactions" which does allow this but no one really implements that besides JanusGraph. will think about this some more but i think that ultimately it will expand scope beyond what this pull request should get into. there is already some documentation on the potential impact to performance that you describe but i will review again.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/Context.java:
```diff
@@ -133,4 +164,28 @@ public void writeAndFlush(final ResponseStatusCode code, final Object responseMe
+        else if (gremlinArgument instanceof String)
+            return RequestContentType.SCRIPT;
+        else
+            return RequestContentType.UNKNOWN;
```
I thought i'd leave it to the calling method to decide what to do with UNKNOWN.

divijvaidya replied:
ok, I didn't check the code but is our calling method handling the case of UNKNOWN?

spmallette replied:
well, note it's a `private` method and its only called in the constructor. the only thing that uses it right now is the code below in `determineTimeout()` and it only needs to check for `SCRIPT`.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/Context.java:
```diff
@@ -45,6 +50,11 @@
     private final GremlinExecutor gremlinExecutor;
     private final ScheduledExecutorService scheduledExecutorService;
     private final AtomicBoolean finalResponseWritten = new AtomicBoolean();
+    private final long requestTimeout;
```
The value is meant to be the timeout for the context (not what is specifically supplied on the request) and thus takes into account the server default. I'll supply some comments for that.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/Settings.java:
```diff
@@ -196,6 +198,59 @@ public Settings() {
+    /**
+     * The time in milliseconds that a {@link UnifiedChannelizer} session can exist. This value cannot be extended
+     * beyond this value irrespective of the number of requests and their individual timeouts. Requests must complete
+     * within this time frame. The default is 10 minutes.
```
good idea - that will needed further clarification.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiTaskSession.java:
```diff
@@ -0,0 +1,283 @@
+            scriptEngineManager = initializeGremlinExecutor(initialSessionTask).getScriptEngineManager();
+
+        scheduledExecutorService = initialSessionTask.getScheduledExecutorService();
+        submitTask(initialSessionTask);
```
smart - more bulletproof that way.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiTaskSession.java:
```diff
@@ -0,0 +1,283 @@
+                while (true) {
+                    // schedule timeout for the current request from the queue
+                    final long seto = sessionTask.getRequestTimeout();
+                    requestCancelFuture = scheduledExecutorService.schedule(
```
the UnifiedHandler starts the timeout for the lifetime of the session. this is the timeout for each request that arrives in the queue.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/UnifiedHandler.java:
```diff
@@ -0,0 +1,327 @@
+    protected static final Set<String> INVALID_BINDINGS_KEYS = new HashSet<>();
+
+    static {
+        INVALID_BINDINGS_KEYS.addAll(Arrays.asList(
```
I dont think we have something exactly like that, but one of the reasons I think we had it this way was so that providers could add their own keys if they extended the class. maybe not the best way to do that do with a mutable set :/

divijvaidya replied:
ok never mind the comment then. resolving this.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/UnifiedHandler.java:
```diff
@@ -0,0 +1,327 @@
+            }
+
+            final Optional<String> optMultiTaskSession = msg.optionalArgs(Tokens.ARGS_SESSION);
+            final String sessionId = optMultiTaskSession.orElse(UUID.randomUUID().toString());
```
ok - i dont think that is a problem.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/UnifiedHandler.java:
```diff
@@ -0,0 +1,327 @@
+                }
+
+                // if the session is done accepting tasks then error time
+                if (!session.submitTask(sessionTask)) {
```
hmm - i suppose i went to too far when i refactored this from the last round. let me add `isAcceptingTasks()` as the code should be general to `Session` here. good catch.

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/UnifiedHandler.java:
```diff
@@ -0,0 +1,327 @@
+                // for a multi or technically a request timeout for a single.
+                if (seto > 0) {
+                    final ScheduledFuture<?> sessionCancelFuture =
+                            scheduledExecutorService.schedule(
```
it is cancelled in `AbstractSession#close()` - adding some javadoc/comments

spmallette commented on gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/UnifiedHandler.java:
```diff
@@ -0,0 +1,327 @@
+                    .statusMessage("Rate limiting").create();
+            ctx.writeAndFlush(response);
+        } finally {
+            ReferenceCountUtil.release(msg);
```
hmm - i think i just copy/pasted from `ExecutorHandler` there. is there a need to `retain()`? perhaps i can just remove the `finally` all together?

DarthRevanXX commented:
hello, when connecting to TinkerPop's Gremlin Server via websockets uses HTTP as the initial transport mechanism, HTTP/1.1 or HTTP/2 ?
